src/modules/hip/kernel/crop_and_patch.hpp

```diff
@@ -0,0 +1,475 @@
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    if ((id_y >= roiTensorPtrSrc[id_z].xywhROI.roiHeight) || (id_x >= roiTensorPtrSrc[id_z].xywhROI.roiWidth))
+    {
+        return;
+    }
+
+    // check if the co-ordinates is within the patch region
+    d_float24 dst_f24;
+    bool rowCheck = (id_y >= patchTensorPtrSrc[id_z].xywhROI.xy.y) && (id_y < (patchTensorPtrSrc[id_z].xywhROI.xy.y + cropTensorPtrSrc[id_z].xywhROI.roiHeight));
+    bool colCheck = (id_x >= patchTensorPtrSrc[id_z].xywhROI.xy.x) && (id_x < (patchTensorPtrSrc[id_z].xywhROI.xy.x + cropTensorPtrSrc[id_z].xywhROI.roiWidth));
+    uint patchStart = patchTensorPtrSrc[id_z].xywhROI.xy.x;
+    uint dstIdx = (id_z * dstStridesNH.x) + (id_y * dstStridesNH.y) + id_x * 3;
+    if(rowCheck && colCheck)
+    {
+        uint srcIdx1 = (id_z * srcStridesNH.x) + (id_y - patchTensorPtrSrc[id_z].xywhROI.xy.y + cropTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y + (id_x - patchTensorPtrSrc[id_z].xywhROI.xy.x + cropTensorPtrSrc[id_z].xywhROI.xy.x) * 3;
+        uint patchEnd = patchTensorPtrSrc[id_z].xywhROI.xy.x + cropTensorPtrSrc[id_z].xywhROI.roiWidth;
+        rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(srcPtr1 + srcIdx1, &dst_f24);
+        // to handle the case when loaded data goes beyond the bounds of patch region
+        if((id_x + 8) >= patchEnd)
+        {
+            uint srcIdx2 = (id_z * srcStridesNH.x) + ((id_y + roiTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y) + (patchEnd + roiTensorPtrSrc[id_z].xywhROI.xy.x) * 3;
+            d_float24 temp_f24;
+            rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(srcPtr2 + srcIdx2, &temp_f24);
+            for(uint i = patchEnd - id_x, j = 0; i < 8; i++, j++)
+            {
+                dst_f24.f8[0].f1[i] = temp_f24.f8[0].f1[j];
+                dst_f24.f8[1].f1[i] = temp_f24.f8[1].f1[j];
+                dst_f24.f8[2].f1[i] = temp_f24.f8[2].f1[j];
+            }
+        }
+        rpp_hip_pack_float24_pln3_and_store24_pkd3(dstPtr + dstIdx, &dst_f24);
+    }
+    // to handle the case when loaded data goes beyond the input region and enters the patch region
+    else if(rowCheck && (id_x + 8) >= patchStart)
+    {
+        uint srcIdx2 = (id_z * srcStridesNH.x) + ((id_y + roiTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y) + (id_x + roiTensorPtrSrc[id_z].xywhROI.xy.x) * 3;
+        rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(srcPtr2 + srcIdx2, &dst_f24);
+        uint srcIdx1 = (id_z * srcStridesNH.x) + ((id_y - patchTensorPtrSrc[id_z].xywhROI.xy.y + cropTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y) + cropTensorPtrSrc[id_z].xywhROI.xy.x * 3;
+        d_float24 temp_f24;
+        rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(srcPtr1 + srcIdx1, &temp_f24);
+        uint patchBegin = patchStart - id_x;
+        for(uint i = patchBegin, j = 0; i < 8; i++, j++)
```

Comment:
- uint patchBegin initialization could be removed.
- patchStart - id_x could directly be used in the for loop like on L91 with for(uint i = patchEnd - id_x......)
- A small device helper for the for loop with signature - (uint patchLoc, d_float24 *dst_f24, d_float24 *temp_f24)
- L98 and L115 for stores are common, could be moved below the if.

Reply:
Just added a separate variable patchStart, since i felt it will make it more readable
Sure will remove it